### DISCUSSION TITLE
Does is support uploading files

### DISCUSSION THREAD

**Discussion by rajeshgautamphd** [Skyvern-AI/skyvern#175] [Q&A] [answered]
I see it supports pop-ups. Does it support uploading files, as that is a regular use case while filling our forms etc. 
can it be done in current version?

**Comment by ykeremy** [comment, marked as answer]
Hello! Yes, Skyvern can upload files to file upload fields. You just need to pass a publicly available link within the navigation payload so that it downloads the file and uploads it to the upload file field. Let us know if you bump into any issues.

**Comment by gwpl** [comment]
> You just need to pass a publicly available link within the navigation payload so that it downloads the file and uploads it to the upload file field. 

Does this advice apply to "upload" block  in workflows ?  

![2024-11-16--10-19-39_Screenshot](https://github.com/user-attachments/assets/22d61d91-ac6a-4217-9578-83ff339db9dd)

There is also "for loop" element.

Two potential use-cases I considered trying it:

* uploading multiple files one by one to some form
* uploading one file to multiple places

So I also wondered how those two different use-cases should be optimally expressed using current workflow building blocks and prompt (if it's too much off topic from this thread, then we can start new "Uploading File + For Loop blocks combined" or sth like that :) )

**Reply by suchintan** [reply]
No -- the upload block just saves any downloaded files to S3 for you!

Uploading files to file selectors is pretty straightforward -- just pass the URLs in navigation payload and skyvern will take care of it for you

We haven't tested uploading multiple files before, but id be curious to see if it works

**Comment by gwpl** [comment]
> No -- the upload block just saves any downloaded files to S3 for you!

So what Download Block is doing then?

![2024-11-16--10-55-19_Screenshot](https://github.com/user-attachments/assets/33f44eef-4211-4a04-abdc-be808b206134)

**Reply by suchintan** [reply]
https://github.com/Skyvern-AI/skyvern/blob/8c653b461a84e392a56dd774fcd2fe0556da82f8/skyvern/forge/sdk/workflow/models/block.py#L788

This downloads a file and saves it to S3 -- it's a subtle and poorly documented nuance:

1. UploadToS3Block: Uploads _all_ downloaded files to S3
2. DownloadBlock: Downlaods a file from the specified URL and uploads it to S3 (for further use in other workflow blocks)

**Reply by gwpl** [reply]
Hmm, apart from documentation maybe renaming to more descriptive names would be helpful?

Brainstorming:

So far it sounds like both are "download" boxes in different ways? (so "upload" is confusing in name...?)

Especially as under button it is written "download from s3", while you just wrote "Downlaods a file from the specified URL and uploads it to S3" which would actually make both buttons semantically different kind of "DownloadToS3" button (?).

Also, does it have to be s3 bucket, even in local deployment?
